core/src/apps/cardano/helpers/cbor_hash_builder.py

```diff
@@ -0,0 +1,62 @@
+from apps.cardano.helpers.lazy_cbor_collection import LazyCborCollection, PauseIteration
+from apps.common import cbor
+
+if False:
+    from typing import Any, Protocol
+
+    class HashFn(Protocol):
+        def update(self, chunk: bytes) -> None:
+            ...
+
+        def digest(self) -> bytes:
+            ...
+
+
+class CborHashBuilder:
+    def __init__(self, hash_fn: HashFn, initial_active_item: LazyCborCollection):
+        self.hash_fn = hash_fn
+        self.active_items_stack: list[LazyCborCollection] = [initial_active_item]
+        self.serializer = cbor.encode_streamed(initial_active_item)
+
+        self.update_hash()
+
+    def update_hash(self) -> None:
+        while (chunk := next(self.serializer)) is not PauseIteration:
+            self.hash_fn.update(chunk)
+
+    def get_hash(self) -> bytes:
+        assert (
+            len(self.active_items_stack) == 1 and self.active_items_stack[0].is_filled()
+        ), "The hash calculation is not finished yet"
+
+        return self.hash_fn.digest()
+
+    def start_lazy_collection(self, collection: LazyCborCollection) -> None:
```

Comment:
The name is somewhat deceptive. What about `add_lazy_collection_to_current_collection`? Maybe not, but a docstring should say that (and "start" is inappropriate, I'd go for `add_lazy_collection`), and don't we want to assert that the current collection is a LazyCborList?

Reply:
Assertion and comment added and renamed to `add_lazy_collection`.